core/tee/tee_enc_fs_key_manager.c

```diff
@@ -0,0 +1,362 @@
+	if (res != TEE_SUCCESS)
+		return res;
+
+	ctx = malloc(hash_ctx_size);
+	if (ctx == NULL)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = crypto_ops.mac.init(ctx, ENC_FS_HMAC_ALG, huk, huk_size);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.mac.update(ctx, ENC_FS_HMAC_ALG,
+			message, message_size);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.hash.final(ctx, ENC_FS_HMAC_ALG, ssk, ssk_size);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = TEE_SUCCESS;
+
+exit:
+	free(ctx);
+	return res;
+}
+
+static TEE_Result enc_fs_init_key_manager(void)
+{
+	int res = TEE_SUCCESS;
+
+	if (enc_fs_ssk.is_init == 0) {
+		struct tee_hw_unique_key huk;
+		uint8_t chip_id[CHIP_ID_LENGTH];
+		uint8_t *message;
+
+		/* Secure Storage Key Generation:
+		 *
+		 *     SSK = HMAC(HUK, message)
+		 *     message := concatenate(chip_id, static string)
+		 * */
+
+		message = malloc(sizeof(chip_id) + sizeof(string_for_ssk_gen));
+		if (message == NULL) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit;
+		}
+
+		tee_otp_get_hw_unique_key(&huk);
+		tee_otp_get_die_id(chip_id, sizeof(chip_id));
+
+		memcpy(message, chip_id, sizeof(chip_id));
+		memcpy(message + sizeof(chip_id), string_for_ssk_gen,
+				sizeof(string_for_ssk_gen));
+
+		res = generate_ssk(
+					enc_fs_ssk.key, sizeof(enc_fs_ssk.key),
+					huk.data, sizeof(huk.data),
+					message, sizeof(message));
+
+		free(message);
+
+		if (res == TEE_SUCCESS)
+			enc_fs_ssk.is_init = 1;
+	}
+
+exit:
+	return res;
+}
+
+static TEE_Result do_file_encryption(struct enc_fs_file_header *hdr,
+			uint8_t *fek, int fek_len,
+			uint8_t *data_in, size_t in_size,
+			uint8_t *data_out, size_t *out_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t *ctx = NULL;
+	size_t ctx_size;
+	uint8_t *dst_data = data_out + sizeof(struct enc_fs_file_header);
+	size_t dst_len = *out_size - sizeof(struct enc_fs_file_header);
+	size_t tag_len = ENC_FS_MAX_TAG_LEN;
+
+	res = crypto_ops.authenc.get_ctx_size(ENC_FS_AUTH_ENC_ALG, &ctx_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	ctx = malloc(ctx_size);
+	if (ctx == NULL)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = crypto_ops.authenc.init(ctx, ENC_FS_AUTH_ENC_ALG,
+			TEE_MODE_ENCRYPT, fek, fek_len, hdr->aad.iv,
+			ENC_FS_IV_LEN, ENC_FS_MAX_TAG_LEN,
+			sizeof(struct aad), in_size);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.update_aad(ctx, ENC_FS_AUTH_ENC_ALG,
+			TEE_MODE_ENCRYPT, (uint8_t *)&hdr->aad,
+			sizeof(struct aad));
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.enc_final(ctx, ENC_FS_AUTH_ENC_ALG,
+			data_in, in_size, dst_data, &dst_len,
+			hdr->tag, &tag_len);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	crypto_ops.authenc.final(ctx, ENC_FS_AUTH_ENC_ALG);
+
+	memcpy(data_out, hdr, sizeof(*hdr));
+	*out_size = sizeof(*hdr) + dst_len;
+
+exit:
+	free(ctx);
+	return res;
+}
+
+static TEE_Result do_file_decryption(struct enc_fs_file_header *hdr,
```

Comment:
It is possible to merge the two function, but it will makes the code flow unclear (need to add some of extra if...else... statement). I suggest to keep the two functions separately.